// polymer-analyzer/lib/polymer/declaration-property-handlers.js
/**
 * @license
 * Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const astValue = require("../javascript/ast-value");
const warning_1 = require("../warning/warning");
const analyze_properties_1 = require("./analyze-properties");
function getBehaviorAssignmentOrWarning(argNode, document) {
    const behaviorName = astValue.getIdentifierName(argNode);
    if (!behaviorName) {
        return {
            kind: 'warning',
            warning: {
                code: 'could-not-determine-behavior-name',
                message: `Could not determine behavior name from expression of type ${argNode
                    .type}`,
                severity: warning_1.Severity.WARNING,
                sourceRange: document.sourceRangeForNode(argNode)
            }
        };
    }
    return {
        kind: 'behaviorAssignment',
        assignment: {
            name: behaviorName,
            sourceRange: document.sourceRangeForNode(argNode)
        }
    };
}
exports.getBehaviorAssignmentOrWarning = getBehaviorAssignmentOrWarning;
/**
 * Returns an object containing functions that will annotate `declaration` with
 * the polymer-specific meaning of the value nodes for the named properties.
 */
function declarationPropertyHandlers(declaration, document) {
    return {
        is(node) {
            if (node.type === 'Literal') {
                declaration.tagName = '' + node.value;
            }
        },
        properties(node) {
            for (const prop of analyze_properties_1.analyzeProperties(node, document)) {
                declaration.addProperty(prop);
            }
        },
        behaviors(node) {
            if (node.type !== 'ArrayExpression') {
                return;
            }
            for (const element of node.elements) {
                const result = getBehaviorAssignmentOrWarning(element, document);
                if (result.kind === 'warning') {
                    declaration.warnings.push(result.warning);
                }
                else {
                    declaration.behaviorAssignments.push(result.assignment);
                }
            }
        },
        observers(node) {
            if (node.type !== 'ArrayExpression') {
                return;
            }
            for (const element of node.elements) {
                let v = astValue.expressionToValue(element);
                if (v === undefined) {
                    v = astValue.CANT_CONVERT;
                }
                declaration.observers.push({ javascriptNode: element, expression: v });
            }
        },
        listeners(node) {
            if (node.type !== 'ObjectExpression') {
                declaration.warnings.push({
                    code: 'invalid-listeners-declaration',
                    message: '`listeners` property should be an object expression',
                    severity: warning_1.Severity.ERROR,
                    sourceRange: document.sourceRangeForNode(node)
                });
                return;
            }
            for (const p of node.properties) {
                const evtName = p.key.type === 'Literal' && p.key.value ||
                    p.key.type === 'Identifier' && p.key.name;
                const handler = p.value.type !== 'Literal' || p.value.value;
                if (typeof evtName !== 'string' || typeof handler !== 'string') {
                    // TODO (maklesoft): Notifiy the user somehow that a listener entry
                    // was not extracted
                    // because the event or handler namecould not be statically analyzed.
                    // E.g. add a low-severity
                    // warning once opting out of rules is supported.
                    continue;
                }
                declaration.listeners.push({ event: evtName, handler: handler });
            }
        }
    };
}
exports.declarationPropertyHandlers = declarationPropertyHandlers;
