// polymer-analyzer/lib/polymer/dom-module-scanner.js
/**
 * @license
 * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator.throw(value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const dom5 = require('dom5');
const parse5_1 = require('parse5');
const model_1 = require('../model/model');
const p = dom5.predicates;
const isDomModule = p.hasTagName('dom-module');
class ScannedDomModule {
    constructor(id, node, sourceRange, ast, slots) {
        this.warnings = [];
        this.id = id;
        this.node = node;
        this.comment = model_1.getAttachedCommentText(node);
        this.sourceRange = sourceRange;
        this.astNode = ast;
        this.slots = slots;
    }
    resolve() {
        return new DomModule(this.node, this.id, this.comment, this.sourceRange, this.astNode, this.warnings, this.slots);
    }
}
exports.ScannedDomModule = ScannedDomModule;
class DomModule {
    constructor(node, id, comment, sourceRange, ast, warnings, slots) {
        this.kinds = new Set(['dom-module']);
        this.identifiers = new Set();
        this.node = node;
        this.id = id;
        this.comment = comment;
        if (id) {
            this.identifiers.add(id);
        }
        this.sourceRange = sourceRange;
        this.astNode = ast;
        this.warnings = warnings;
        this.slots = slots;
    }
}
exports.DomModule = DomModule;
class DomModuleScanner {
    scan(document, visit) {
        return __awaiter(this, void 0, void 0, function* () {
            let domModules = [];
            yield visit((node) => {
                if (isDomModule(node)) {
                    const template = dom5.query(node, dom5.predicates.hasTagName('template'));
                    let slots = [];
                    if (template) {
                        slots = dom5.queryAll(parse5_1.treeAdapters.default.getTemplateContent(template), dom5.predicates.hasTagName('slot')).map(s => new model_1.Slot(dom5.getAttribute(s, 'name') || '', document.sourceRangeForNode(s)));
                    }
                    domModules.push(new ScannedDomModule(dom5.getAttribute(node, 'id'), node, document.sourceRangeForNode(node), node, slots));
                }
            });
            return domModules;
        });
    }
}
exports.DomModuleScanner = DomModuleScanner;
