// polymer-analyzer/lib/polymer/js-utils.js
/**
 * @license
 * Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const escodegen = require("escodegen");
const esutil_1 = require("../javascript/esutil");
const jsdoc = require("../javascript/jsdoc");
const warning_1 = require("../warning/warning");
/**
 * Create a ScannedProperty object from an estree Property AST node.
 */
function toScannedPolymerProperty(node, sourceRange) {
    const type = esutil_1.closureType(node.value, sourceRange);
    const parsedJsdoc = jsdoc.parseJsdoc(esutil_1.getAttachedComment(node) || '');
    const description = jsdoc.removeLeadingAsterisks(esutil_1.getAttachedComment(node) || '').trim();
    const maybeName = esutil_1.objectKeyToString(node.key);
    const warnings = [];
    if (!maybeName) {
        warnings.push({
            code: 'unknown-prop-name',
            message: `Could not determine name of property from expression of type: ${node
                .key.type}`,
            sourceRange: sourceRange,
            severity: warning_1.Severity.WARNING
        });
    }
    const name = maybeName || '';
    const result = {
        name,
        type,
        description,
        sourceRange,
        warnings,
        astNode: node,
        isConfiguration: configurationProperties.has(name),
        jsdoc: parsedJsdoc,
        privacy: getOrInferPrivacy(name, parsedJsdoc, false)
    };
    return result;
}
exports.toScannedPolymerProperty = toScannedPolymerProperty;
;
/**
 * Properties on Polymer element prototypes that are part of Polymer's
 * configuration syntax.
 */
const configurationProperties = new Set([
    'attached',
    'attributeChanged',
    'beforeRegister',
    'configure',
    'constructor',
    'created',
    'detached',
    'enableCustomStyleProperties',
    'extends',
    'hostAttributes',
    'is',
    'listeners',
    'mixins',
    'observers',
    'properties',
    'ready',
    'registered',
]);
/**
 * Create a ScannedMethod object from an estree Property AST node.
 */
function toScannedMethod(node, sourceRange) {
    const scannedMethod = toScannedPolymerProperty(node, sourceRange);
    if (scannedMethod.type === 'Function' ||
        scannedMethod.type === 'ArrowFunction') {
        const value = node.value;
        scannedMethod.params = (value.params || []).map((param) => {
            // With ES6 we can have a lot of param patterns. Best to leave the
            // formatting to escodegen.
            return { name: escodegen.generate(param) };
        });
    }
    return scannedMethod;
}
exports.toScannedMethod = toScannedMethod;
function getOrInferPrivacy(name, annotation, privateUnlessDocumented) {
    const explicitPrivacy = jsdoc.getPrivacy(annotation);
    const specificName = name.slice(name.lastIndexOf('.') + 1);
    if (explicitPrivacy) {
        return explicitPrivacy;
    }
    else if (specificName.startsWith('__')) {
        return 'private';
    }
    else if (specificName.startsWith('_')) {
        return 'protected';
    }
    else if (specificName.endsWith('_')) {
        return 'private';
    }
    else if (configurationProperties.has(specificName)) {
        return 'protected';
    }
    else {
        if (privateUnlessDocumented) {
            // Some members, like methods or properties on classes are private by
            // default unless they have documentation.
            const hasDocs = !!annotation && !jsdoc.isAnnotationEmpty(annotation);
            return hasDocs ? 'public' : 'private';
        }
        else {
            // Other members, like entries in the Polymer `properties` block are
            // public unless there are clear signals otherwise.
            return 'public';
        }
    }
}
exports.getOrInferPrivacy = getOrInferPrivacy;
