// polymer-analyzer/lib/polymer/polymer-element-mixin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const model_1 = require("../model/model");
const js_utils_1 = require("./js-utils");
const polymer_element_1 = require("./polymer-element");
class ScannedPolymerElementMixin extends model_1.ScannedElementMixin {
    constructor(options) {
        super();
        this.properties = [];
        this.methods = [];
        this.observers = [];
        this.listeners = [];
        this.behaviorAssignments = [];
        // FIXME(rictic): domModule and scriptElement aren't known at a file local
        //     level. Remove them here, they should only exist on PolymerElement.
        this.domModule = undefined;
        this.scriptElement = undefined;
        // Indicates if an element is a pseudo element
        this.pseudo = false;
        this.abstract = false;
        // TODO(justinfagnani): fix this constructor to not be crazy, or remove
        // class altogether.
        const optionsCopy = Object.assign({}, options);
        Object.assign(this, optionsCopy);
    }
    addProperty(prop) {
        polymer_element_1.addProperty(this, prop);
    }
    addMethod(method) {
        polymer_element_1.addMethod(this, method);
    }
    resolve(document) {
        const element = new PolymerElementMixin();
        Object.assign(element, this);
        for (const method of element.methods) {
            // methods are only public by default if they're documented.
            method.privacy = js_utils_1.getOrInferPrivacy(method.name, method.jsdoc, true);
        }
        element.mixins = [];
        for (const mixin of this.mixins) {
            // TODO(rictic): we should mix these mixins into `this`. See
            // PolymerElement's logic for this.
            element.mixins.push(mixin.resolve(document));
        }
        return element;
    }
}
exports.ScannedPolymerElementMixin = ScannedPolymerElementMixin;
class PolymerElementMixin extends model_1.ElementMixin {
    constructor() {
        super();
        this.localIds = [];
        this.kinds = new Set(['element-mixin', 'polymer-element-mixin']);
        this.behaviorAssignments = [];
    }
    emitPropertyMetadata(property) {
        const polymerMetadata = {};
        const polymerMetadataFields = ['notify', 'observer', 'readOnly'];
        for (const field of polymerMetadataFields) {
            if (field in property) {
                polymerMetadata[field] = property[field];
            }
        }
        return { polymer: polymerMetadata };
    }
}
exports.PolymerElementMixin = PolymerElementMixin;
