// polymer-analyzer/lib/model/import.js
/**
 * @license
 * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
"use strict";
/**
 * Represents an import, such as an HTML import, an external script or style
 * tag, or an JavaScript import.
 *
 * @template N The AST node type
 */
class ScannedImport {
    constructor(type, url, sourceRange, urlSourceRange, ast) {
        this.warnings = [];
        this.type = type;
        this.url = url;
        this.sourceRange = sourceRange;
        this.urlSourceRange = urlSourceRange;
        this.astNode = ast;
    }
    resolve(document) {
        const importedDocument = document.analyzer._getDocument(this.url);
        return importedDocument && new Import(this.url, this.type, importedDocument, this.sourceRange, this.urlSourceRange, this.astNode, this.warnings);
    }
}
exports.ScannedImport = ScannedImport;
class Import {
    constructor(url, type, document, sourceRange, urlSourceRange, ast, warnings) {
        this.identifiers = new Set();
        this.kinds = new Set(['import']);
        this.url = url;
        this.type = type;
        this.document = document;
        this.kinds.add(this.type);
        this.sourceRange = sourceRange;
        this.urlSourceRange = urlSourceRange;
        this.astNode = ast;
        this.warnings = warnings;
    }
    toString() {
        return `<Import type="${this.type}" url="${this.url}">`;
    }
}
exports.Import = Import;
