// polymer-analyzer/lib/model/package.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// A regexp that matches paths to external code.
// TODO(rictic): Make this extensible (polymer.json?).
const MATCHES_EXTERNAL = /(^|\/)(bower_components|node_modules)/;
/**
 * Represents a queryable interface over all documents in a package/project.
 *
 * Results of queries will include results from all documents in the package, as
 * well as from external dependencies that are transitively imported by
 * documents in the package.
 */
class Package {
    static isExternal(path) {
        return MATCHES_EXTERNAL.test(path);
    }
    constructor(documents, warnings) {
        const potentialRoots = new Set(documents);
        this._toplevelWarnings = warnings;
        // We trim down the set of documents as a performance optimization. We only
        // need a set of documents such that all other documents we're interested in
        // can be reached from them. That way we'll do less duplicate work when we
        // query over all documents.
        for (const doc of potentialRoots) {
            for (const imprt of doc.getByKind('import', { imported: true })) {
                // When there's cycles we can keep any element of the cycle, so why not
                // this one.
                if (imprt.document !== doc) {
                    potentialRoots.delete(imprt.document);
                }
            }
        }
        this._documents = potentialRoots;
    }
    getByKind(kind, options) {
        const result = new Set();
        const docQueryOptions = this._getDocumentQueryOptions(options);
        for (const doc of this._documents) {
            addAll(result, doc.getByKind(kind, docQueryOptions));
        }
        return result;
    }
    getById(kind, identifier, options) {
        const result = new Set();
        const docQueryOptions = this._getDocumentQueryOptions(options);
        for (const doc of this._documents) {
            addAll(result, doc.getById(kind, identifier, docQueryOptions));
        }
        return result;
    }
    getOnlyAtId(kind, identifier, options) {
        const results = this.getById(kind, identifier, options);
        if (results.size > 1) {
            throw new Error(`Expected to find at most one ${kind} with id ${identifier} ` +
                `but found ${results.size}.`);
        }
        ;
        return results.values().next().value || undefined;
    }
    /**
     * Get all features for all documents in the project or their imports.
     */
    getFeatures(options) {
        const result = new Set();
        const docQueryOptions = this._getDocumentQueryOptions(options);
        for (const doc of this._documents) {
            addAll(result, doc.getFeatures(docQueryOptions));
        }
        return result;
    }
    /**
     * Get all warnings in the project.
     */
    getWarnings(options) {
        const result = new Set(this._toplevelWarnings);
        const docQueryOptions = this._getDocumentQueryOptions(options);
        for (const doc of this._documents) {
            addAll(result, new Set(doc.getWarnings(docQueryOptions)));
        }
        return Array.from(result);
    }
    _getDocumentQueryOptions(options) {
        return {
            imported: true,
            externalPackages: !!(options && options.externalPackages)
        };
    }
}
exports.Package = Package;
function addAll(set1, set2) {
    for (const val of set2) {
        set1.add(val);
    }
    return set1;
}
