// polymer-analyzer/lib/model/source-range.js
/**
 * @license
 * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Corrects source ranges based on an offset.
 *
 * Source ranges for inline documents need to be corrected relative to their
 * positions in their containing documents.
 *
 * For example, if a <script> tag appears in the fifth line of its containing
 * document, we need to move all the source ranges inside that script tag down
 * by 5 lines. We also need to correct the column offsets, but only for the
 * first line of the <script> contents.
 */
function correctSourceRange(sourceRange, locationOffset) {
    if (!locationOffset || !sourceRange) {
        return sourceRange;
    }
    return {
        file: locationOffset.filename || sourceRange.file,
        start: correctPosition(sourceRange.start, locationOffset),
        end: correctPosition(sourceRange.end, locationOffset)
    };
}
exports.correctSourceRange = correctSourceRange;
function correctPosition(position, locationOffset) {
    return {
        line: position.line + locationOffset.line,
        column: position.column + (position.line === 0 ? locationOffset.col : 0)
    };
}
exports.correctPosition = correctPosition;
