// polymer-analyzer/lib/test/polymer/polymer2-element-scanner_test.js
/**
 * @license
 * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const path = require("path");
const javascript_parser_1 = require("../../javascript/javascript-parser");
const polymer_element_1 = require("../../polymer/polymer-element");
const polymer2_element_scanner_1 = require("../../polymer/polymer2-element-scanner");
const fs_url_loader_1 = require("../../url-loader/fs-url-loader");
const test_utils_1 = require("../test-utils");
chai_1.use(require('chai-subset'));
suite('Polymer2ElementScanner', () => {
    const testFilesDir = path.resolve(__dirname, '../static/polymer2/');
    const urlLoader = new fs_url_loader_1.FSUrlLoader(testFilesDir);
    const underliner = new test_utils_1.CodeUnderliner(urlLoader);
    function getElements(filename) {
        return __awaiter(this, void 0, void 0, function* () {
            const file = yield urlLoader.load(filename);
            const parser = new javascript_parser_1.JavaScriptParser();
            const document = parser.parse(file, filename);
            const scanner = new polymer2_element_scanner_1.Polymer2ElementScanner();
            const visit = (visitor) => Promise.resolve(document.visit([visitor]));
            const features = yield scanner.scan(document, visit);
            return features.filter((e) => e instanceof polymer_element_1.ScannedPolymerElement);
        });
    }
    ;
    function getTestProps(element) {
        const props = {
            className: element.className,
            superClass: element.superClass && element.superClass.identifier,
            tagName: element.tagName,
            description: element.description,
            summary: element.summary,
            properties: element.properties.map((p) => ({
                name: p.name,
            })),
            attributes: element.attributes.map((a) => ({
                name: a.name,
            })),
            methods: element.methods.map((m) => ({ name: m.name, params: m.params, return: m.return })),
        };
        if (element.mixins.length > 0) {
            props.mixins = element.mixins.map((m) => m.identifier);
        }
        return props;
    }
    test('Finds two basic elements', () => __awaiter(this, void 0, void 0, function* () {
        const elements = yield getElements('test-element-1.js');
        const elementData = elements.map(getTestProps);
        chai_1.assert.deepEqual(elementData, [
            {
                tagName: 'test-element',
                className: 'TestElement',
                superClass: 'Polymer.Element',
                description: '',
                summary: '',
                properties: [{
                        name: 'foo',
                    }],
                attributes: [{
                        name: 'foo',
                    }],
                methods: [],
            },
            {
                tagName: undefined,
                className: 'BaseElement',
                superClass: 'Polymer.Element',
                description: 'A very basic element',
                summary: 'A basic element',
                properties: [{
                        name: 'foo',
                    }],
                attributes: [{
                        name: 'foo',
                    }],
                methods: [],
            },
        ]);
        const underlinedSource1 = yield underliner.underline(elements[0].sourceRange);
        chai_1.assert.equal(underlinedSource1, `
class TestElement extends Polymer.Element {
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  static get properties() {
~~~~~~~~~~~~~~~~~~~~~~~~~~~
    return {
~~~~~~~~~~~~
      foo: {
~~~~~~~~~~~~
        notify: true,
~~~~~~~~~~~~~~~~~~~~~
        type: String,
~~~~~~~~~~~~~~~~~~~~~
      }
~~~~~~~
    }
~~~~~
  }
~~~
}
~`);
        const underlinedSource2 = yield underliner.underline(elements[1].sourceRange);
        chai_1.assert.equal(underlinedSource2, `
class BaseElement extends Polymer.Element {
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  static get properties() {
~~~~~~~~~~~~~~~~~~~~~~~~~~~
    return {
~~~~~~~~~~~~
      foo: {
~~~~~~~~~~~~
        notify: true,
~~~~~~~~~~~~~~~~~~~~~
        type: String,
~~~~~~~~~~~~~~~~~~~~~
      },
~~~~~~~~
    };
~~~~~~
  }
~~~
}
~`);
    }));
    test('Uses static is getter for tagName', () => __awaiter(this, void 0, void 0, function* () {
        const elements = yield getElements('test-element-2.js');
        const elementData = elements.map(getTestProps);
        chai_1.assert.deepEqual(elementData, [
            {
                tagName: 'test-element',
                className: 'TestElement',
                superClass: 'HTMLElement',
                description: '',
                summary: '',
                properties: [],
                attributes: [],
                methods: [],
            },
        ]);
    }));
    test('Finds vanilla elements', () => __awaiter(this, void 0, void 0, function* () {
        const elements = yield getElements('test-element-4.js');
        const elementData = elements.map(getTestProps);
        chai_1.assert.deepEqual(elementData, [
            {
                tagName: 'test-element',
                className: 'TestElement',
                superClass: 'HTMLElement',
                description: '',
                summary: '',
                properties: [],
                attributes: [
                    {
                        name: 'a',
                    },
                    {
                        name: 'b',
                    }
                ],
                methods: [],
            },
        ]);
    }));
    test('Observed attributes override induced attributes', () => __awaiter(this, void 0, void 0, function* () {
        const elements = yield getElements('test-element-5.js');
        const elementData = elements.map(getTestProps);
        chai_1.assert.deepEqual(elementData, [
            {
                tagName: 'test-element',
                className: 'TestElement',
                superClass: 'Polymer.Element',
                description: '',
                summary: '',
                properties: [{
                        name: 'foo',
                    }],
                attributes: [
                    {
                        name: 'a',
                    },
                    {
                        name: 'b',
                    }
                ],
                methods: [],
            },
        ]);
    }));
    test('properly sets className for elements with the memberof tag', () => __awaiter(this, void 0, void 0, function* () {
        const elements = yield getElements('test-element-8.js');
        const elementData = elements.map(getTestProps);
        chai_1.assert.deepEqual(elementData, [
            {
                tagName: 'test-element-one',
                className: 'Polymer.TestElementOne',
                superClass: 'Polymer.Element',
                description: `This element is a member of Polymer namespace and is registered with its
namespaced name.`,
                summary: '',
                properties: [{
                        name: 'foo',
                    }],
                attributes: [{
                        name: 'foo',
                    }],
                methods: [],
            },
            {
                tagName: 'test-element-two',
                className: 'Polymer.TestElementTwo',
                superClass: 'Polymer.Element',
                description: `This element is a member of Polymer namespace and is registered without its
namespaced name.`,
                summary: '',
                properties: [{
                        name: 'foo',
                    }],
                attributes: [{
                        name: 'foo',
                    }],
                methods: [],
            }
        ]);
    }));
    test('Read mixes annotations', () => __awaiter(this, void 0, void 0, function* () {
        const elements = yield getElements('test-element-6.js');
        const elementData = elements.map(getTestProps);
        chai_1.assert.deepEqual(elementData, [
            {
                tagName: 'test-element',
                className: 'TestElement',
                superClass: 'Polymer.Element',
                description: '',
                summary: '',
                properties: [],
                attributes: [],
                methods: [],
                mixins: ['Mixin2', 'Mixin1'],
            },
        ]);
    }));
    test('Reads just mixin application', () => __awaiter(this, void 0, void 0, function* () {
        const elements = yield getElements('test-element-9.js');
        const elementData = elements.map(getTestProps);
        chai_1.assert.deepEqual(elementData, [
            {
                tagName: undefined,
                className: 'BaseElement',
                superClass: 'Polymer.Element',
                description: '',
                summary: '',
                properties: [],
                attributes: [],
                methods: [],
            },
            {
                tagName: undefined,
                className: 'SubElement',
                superClass: 'BaseElement',
                description: '',
                summary: '',
                properties: [],
                attributes: [],
                methods: [],
                mixins: ['Mixin'],
            },
            {
                tagName: undefined,
                className: 'SubElement2',
                superClass: 'BaseElement',
                description: '',
                summary: '',
                properties: [],
                attributes: [],
                methods: [],
                mixins: ['Mixin'],
            },
        ]);
    }));
    test('properly reads properties and methods of elements and element classes', () => __awaiter(this, void 0, void 0, function* () {
        const elements = yield getElements('test-element-10.js');
        const elementData = elements.map(getTestProps);
        chai_1.assert.deepEqual(elementData, [
            {
                tagName: 'test-element',
                className: 'TestElement',
                superClass: 'Polymer.Element',
                description: ``,
                summary: '',
                properties: [{
                        name: 'foo',
                    }],
                attributes: [{
                        name: 'foo',
                    }],
                methods: [
                    { name: 'customInstanceFunction', params: [], return: undefined },
                    {
                        name: 'customInstanceFunctionWithJSDoc',
                        params: [], return: undefined
                    },
                    {
                        name: 'customInstanceFunctionWithParams',
                        params: [{ name: 'a' }, { name: 'b' }, { name: 'c' }],
                        return: undefined,
                    },
                    {
                        name: 'customInstanceFunctionWithParamsAndJSDoc',
                        params: [{ name: 'a' }, { name: 'b' }, { name: 'c' }],
                        return: undefined,
                    },
                    {
                        name: 'customInstanceFunctionWithParamsAndPrivateJSDoc',
                        params: [], return: undefined,
                    },
                ],
            },
        ]);
    }));
});
