// polymer-analyzer/lib/test/polymer/polymer-element-scanner_test.js
/**
 * @license
 * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/// <reference path="../../../node_modules/@types/mocha/index.d.ts" />
const chai_1 = require("chai");
const javascript_parser_1 = require("../../javascript/javascript-parser");
const polymer_element_scanner_1 = require("../../polymer/polymer-element-scanner");
suite('PolymerElementScanner', () => {
    suite('scan()', () => {
        let scanner;
        setup(() => {
            scanner = new polymer_element_scanner_1.PolymerElementScanner();
        });
        test('finds polymer elements', () => __awaiter(this, void 0, void 0, function* () {
            const contents = `Polymer({
        is: 'x-foo',
        properties: {
          a: {
            type: Boolean,
            value: 5
          },
          b: {
            type: String,
            value: function() {
              return 'test';
            }
          },
          c: {
            type: Number,
            readOnly: true
          },
          d: {
            type: Number,
            computed: '_computeD()'
          },
          e: {
            type: String,
            notify: true
          },
          f: {
            type: Object,
            observer: '_observeF'
          },
          g: {
            type: {},
            computed: '_computeG()',
            readOnly: false
          },
          h: String,
          all: {
            type: Object,
            notify: true,
            readOnly: false,
            reflectToAttribute: false,
            observer: '_observeAll'
          }
        },
        observers: [
          '_anObserver()',
          '_anotherObserver()'
        ],
        listeners: {
          'event-a': '_handleA',
          eventb: '_handleB',
          'event-c': _handleC,
          [['event', 'd'].join('-')]: '_handleD'
        },
        customPublicMethod: (foo, bar) => { return foo + bar; },
        _customPrivateMethod: (foo, bar) => { return foo + bar; },
        /** This is an instance method with JS Doc */
        customPublicMethodWithJsDoc: (foo, bar) => { return foo + bar; },
        customPublicMethodWithClassicFunction: function(foo, bar) { return foo + bar; },
      });
      Polymer({
        is: 'x-bar',
        listeners: []
      });`;
            const document = new javascript_parser_1.JavaScriptParser().parse(contents, 'test-document.html');
            const visit = (visitor) => __awaiter(this, void 0, void 0, function* () { return document.visit([visitor]); });
            const features = yield scanner.scan(document, visit);
            chai_1.assert.deepEqual(features.map((f) => f.tagName), ['x-foo', 'x-bar']);
            chai_1.assert.deepEqual(features[0].observers.map((o) => o.expression), ['_anObserver()', '_anotherObserver()']);
            chai_1.assert.deepEqual(features[0].events.map((e) => e.name), ['e-changed', 'all-changed']);
            chai_1.assert.equal(features[0].properties.length, 9);
            chai_1.assert.deepEqual(features[0]
                .properties.filter((p) => p.warnings.length > 0)
                .map((p) => [p.name, p.warnings.map((w) => w.message)]), [[
                    'g',
                    [
                        'Invalid type in property object.',
                        'Unable to determine type for property.'
                    ]
                ]]);
            chai_1.assert.deepEqual(features[0].methods.map((m) => m.name), [
                'customPublicMethod',
                '_customPrivateMethod',
                'customPublicMethodWithJsDoc',
                'customPublicMethodWithClassicFunction',
            ]);
            chai_1.assert.deepEqual(features[0].properties.map((p) => [p.name, p.type]), [
                ['a', 'boolean'],
                ['b', 'string'],
                ['c', 'number'],
                ['d', 'number'],
                ['e', 'string'],
                ['f', 'Object'],
                ['g', undefined],
                ['h', 'string'],
                ['all', 'Object']
            ]);
            chai_1.assert.deepEqual(features[0].attributes.map((p) => [p.name, p.changeEvent]), [
                ['a', undefined],
                ['b', undefined],
                ['c', undefined],
                ['d', undefined],
                ['e', 'e-changed'],
                ['f', undefined],
                ['g', undefined],
                ['h', undefined],
                ['all', 'all-changed']
            ]);
            chai_1.assert.deepEqual(features[0].properties.filter((p) => p.readOnly).map((p) => p.name), ['c', 'd', 'g']);
            chai_1.assert.deepEqual(features[0]
                .properties.filter((p) => p.default)
                .map((p) => [p.name, p.default]), [['a', '5'], ['b', '"test"']]);
            chai_1.assert.deepEqual(features[0].properties.filter((p) => p.notify).map((p) => p.name), ['e', 'all']);
            chai_1.assert.deepEqual(features[0].listeners, [
                { event: 'event-a', handler: '_handleA' },
                { event: 'eventb', handler: '_handleB' }
            ]);
            // Skip not statically analizable entries without emitting a warning
            chai_1.assert.equal(features[0]
                .warnings
                .filter((w) => w.code === 'invalid-listeners-declaration')
                .length, 0);
            // Emit warning for non-object `listeners` literal
            chai_1.assert.equal(features[1]
                .warnings
                .filter((w) => w.code === 'invalid-listeners-declaration')
                .length, 1);
        }));
        test('Polymer 2 class observers crash', () => __awaiter(this, void 0, void 0, function* () {
            // When Polymer 2 adopted a static getter for observers, it crashed
            // the Polymer 1 element scanner.
            const contents = `class TestElement extends Polymer.Element {
        static get observers() {
          return foo.bar;
        }
      }`;
            const document = new javascript_parser_1.JavaScriptParser().parse(contents, 'test-document.html');
            const visit = (visitor) => __awaiter(this, void 0, void 0, function* () { return document.visit([visitor]); });
            // Scanning should not throw
            yield scanner.scan(document, visit);
        }));
    });
});
