// polymer-analyzer/lib/editor-service/local-editor-service.js
/**
 * @license
 * Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const analyzer_1 = require("../analyzer");
const html_document_1 = require("../html/html-document");
const warning_1 = require("../warning/warning");
const ast_from_source_position_1 = require("./ast-from-source-position");
const editor_service_1 = require("./editor-service");
class LocalEditorService extends editor_service_1.EditorService {
    constructor(options) {
        super();
        this._analyzer = new analyzer_1.Analyzer(options);
    }
    fileChanged(localPath, contents) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._analyzer.analyze(localPath, contents);
        });
    }
    getDocumentationAtPosition(localPath, position) {
        return __awaiter(this, void 0, void 0, function* () {
            const feature = yield this._getFeatureAt(localPath, position);
            if (!feature) {
                return;
            }
            if (isProperty(feature)) {
                if (feature.type) {
                    return `{${feature.type}} ${feature.description}`;
                }
            }
            return feature.description;
        });
    }
    getDefinitionForFeatureAtPosition(localPath, position) {
        return __awaiter(this, void 0, void 0, function* () {
            const feature = yield this._getFeatureAt(localPath, position);
            if (!feature) {
                return;
            }
            return feature.sourceRange;
        });
    }
    getTypeaheadCompletionsAtPosition(localPath, position) {
        return __awaiter(this, void 0, void 0, function* () {
            const document = yield this._analyzer.analyze(localPath);
            const location = yield this._getLocationResult(document, position);
            if (!location) {
                return;
            }
            if (location.kind === 'tagName' || location.kind === 'text') {
                const elements = Array.from(document.getByKind('element')).filter(e => e.tagName);
                return {
                    kind: 'element-tags',
                    elements: elements.map(e => {
                        const attributesSpace = e.attributes.length > 0 ? ' ' : '';
                        return {
                            tagname: e.tagName,
                            description: e.description,
                            expandTo: location.kind === 'text' ?
                                `<${e.tagName}${attributesSpace}></${e.tagName}>` :
                                undefined
                        };
                    })
                };
            }
            else if (location.kind === 'attribute') {
                const elements = document.getById('element', location.element.nodeName);
                let attributes = [];
                for (const element of elements) {
                    // A map from the inheritedFrom to a sort prefix. Note that
                    // `undefined` is a legal value for inheritedFrom.
                    const sortPrefixes = new Map();
                    // Not inherited, that means local! Sort it early.
                    sortPrefixes.set(undefined, 'aaa-');
                    if (element.superClass) {
                        sortPrefixes.set(element.superClass, 'bbb-');
                    }
                    if (element.extends) {
                        sortPrefixes.set(element.extends, 'ccc-');
                    }
                    const elementAttributes = element.attributes.map(p => {
                        const sortKey = (sortPrefixes.get(p.inheritedFrom) || `ddd-`) + p.name;
                        return {
                            name: p.name,
                            description: p.description || '',
                            type: p.type,
                            inheritedFrom: p.inheritedFrom, sortKey
                        };
                    });
                    const eventAttributes = element.events.map((e) => {
                        const postfix = sortPrefixes.get(e.inheritedFrom) || 'ddd-';
                        const sortKey = `eee-${postfix}on-${e.name}`;
                        return {
                            name: `on-${e.name}`,
                            description: e.description || '',
                            type: e.type || 'CustomEvent',
                            inheritedFrom: e.inheritedFrom, sortKey
                        };
                    });
                    attributes =
                        attributes.concat(elementAttributes).concat(eventAttributes);
                }
                return { kind: 'attributes', attributes };
            }
            ;
        });
    }
    getWarningsForFile(localPath) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const doc = yield this._analyzer.analyze(localPath);
                return doc.getWarnings();
            }
            catch (e) {
                // This might happen if, e.g. `localPath` has a parse error. In that case
                // we can't construct a valid Document, but we might still be able to give
                // a reasonable warning.
                if (e instanceof warning_1.WarningCarryingException) {
                    const warnException = e;
                    return [warnException.warning];
                }
                throw e;
            }
        });
    }
    _clearCaches() {
        return __awaiter(this, void 0, void 0, function* () {
            this._analyzer.clearCaches();
        });
    }
    _getFeatureAt(localPath, position) {
        return __awaiter(this, void 0, void 0, function* () {
            const document = yield this._analyzer.analyze(localPath);
            const location = yield this._getLocationResult(document, position);
            if (!location) {
                return;
            }
            if (location.kind === 'tagName') {
                return document.getOnlyAtId('element', location.element.nodeName);
            }
            else if (location.kind === 'attribute') {
                const elements = document.getById('element', location.element.nodeName);
                if (elements.size === 0) {
                    return;
                }
                return concatMap(elements, (el) => el.attributes)
                    .find(at => at.name === location.attribute);
            }
        });
    }
    _getLocationResult(document, position) {
        return __awaiter(this, void 0, void 0, function* () {
            const parsedDocument = document.parsedDocument;
            if (!(parsedDocument instanceof html_document_1.ParsedHtmlDocument)) {
                return;
            }
            return ast_from_source_position_1.getLocationInfoForPosition(parsedDocument, position);
        });
    }
}
exports.LocalEditorService = LocalEditorService;
function isProperty(d) {
    return 'type' in d;
}
function concatMap(inputs, f) {
    let results = [];
    for (const input of inputs) {
        results = results.concat(f(input));
    }
    return results;
}
